### ISSUE TITLE
Redefine concept of "sandbox" and define interface for creating one

### ISSUE THREAD

**Issue by haircommander** [cri-o/cri-o#5622]
TL;DR: this proposal describes a clean way to integrate the work being done for [conmon 3.0/conmon-rs](https://github.com/containers/conmon-rs), by redefining how CRI-O creates sandboxes/pods

Historically, sandboxes in CRI-O have been very tied to the infra container. Most of the sandbox creation code involves creating the infra container. However, there exists a world (perhaps in the not too distant future!) that the (oci) infra container isn't needed at all! 

Moving towards that world, we'll need to redefine two things. 

####  The specific responsibilities of a sandbox. Specifically, a sandbox should:
  - hold the namespaces for the pod
  - hold the shm path for the pod
  - hold the pod's runtime class [1]
  - hold the information that each container falls back to if not defined in the container spec (selinux label, for instance)
  - hold the user namespace mapping for the pod
  - be able to be restored from disk (currently done by restoring annotations in the config.json, but maybe we can do better...)

Updating the mental model of the responsibilities of a sandbox will help us lead into the bigger piece of work, which is redefining...

#### how a sandbox is created:

I propose we expand the sandbox creation to have 5 total interfaces:
- same kernel process - infra
  - In this, we have a pod whose namespaces are created by pinns, and the infra container is spoofed out, but not actually created in runtime
  - i.e. the current way we do drop_infra_ctr
- same kernel process + infra
  - In this, we have a pod whose namespaces are created by pinns, and the infra container is created
  - i.e. how we currently create a pod with a pod level PID namespace
- different kernel process (runtime VM)
  - In this, the infra container creation is done, and it actually creates a VM.
  - little should change here
- **NEW** same kernel process - infra + conmon-rs
  - In this, we have a pod whose namespaces are created by pinns *or conmon-rs* [2].
  - sandbox creation process will also involve creating conmon-rs instance
  - runtime class instance will refer to conmon-rs
- **NEW** same kernel process + pod PID namespace + conmon-rs
  - In this, we have a pod whose namespaces are created by *conmon-rs* [2].
    - Specifically, we create the pod PID namespace in conmon-rs, who is responsible for holding it open for the lifecycle of the pod.
  - sandbox creation process will also involve creating conmon-rs instance
  - runtime class instance will refer to conmon-rs
  - **no** infra container creation will be done.
  
The net effect will be a way to transition between the conmon 2.0 days (the first three items) to the conmon 3.0 days (the last three items). Something to note is the final two items will likely merge into one once conmon-rs has support to be the holder of the PID namespace.

1: Currently, runtime classes are owned by the server, and found via mapping. I wasn't around for this decision, but I think in the future it will make more sense for the sandbox object to hold the runtime class.
2: It is not in the MVP roadmap for conmon-rs to create pod level namespaces, but it is in the distant roadmap. It's possible the namespaces will be created by CRI-O for a time, before conmon-rs has support. However, to fully drop the infra container in the pod PID namespace case, conmon-rs will have to have the functionality. Ultimately, factoring out namespace creation to a piece of the interface will aid in this transition, regardless of when it makes it into CRI-O

**Comment by saschagrunert** [comment]
A friendly reminder that this issue had no activity for 30 days.
